### ISSUE TITLE
Mixed precision return fp32

### ISSUE THREAD

**Issue by edgarriba** [huggingface/accelerate#382]
hi ! Thanks for this amazing core library :)

 I was trying to make the `Accelerator` to work with the Kornia trainers and I just found that for some reason the data coming from the DataLoader and also the Model, is not casted to half precision. Maybe i'm using the library wrong -- I abstracted a bit to show overall what i do:

```python
import torch
from accelerate import Accelerator

class DummyDataset(torch.utils.data.Dataset):
    def __len__(self):
        return 10

    def __getitem__(self, index):
        return torch.ones(1, 32, 32)

ds = DummyDataset()
dl = torch.utils.data.DataLoader(ds, batch_size=1)
mod = torch.nn.Conv2d(1,1,1)

acc = Accelerator(cpu=False, mixed_precision="fp16")
dl, mod = acc.prepare(dl, mod)

img = next(iter(dl))
out = mod(img)

print(img.shape)
print(img.device)
print(img.dtype)

print(out.shape)
print(out.device)
print(out.dtype)
```
The result of this is:
```bash
torch.Size([1, 1, 32, 32])
cuda:0
torch.float32
torch.Size([1, 1, 32, 32])
cuda:0
torch.float32
```
I was expecting that both the `dl` and `mod`/`out` would output a `torch.float16` dtype.

The colab to reproduce is here: https://colab.research.google.com/drive/1NQJpVUnCoZH-bsPGwsjfD54FqlLcM_rR?usp=sharing

**Comment by sgugger** [comment]
Hi Edgar! Glad you love the library :-)

I think your expectations come from the old days of mixed precision with apex. With the new PyTorch AMP, and following [their tutorial](https://pytorch.org/docs/stable/notes/amp_examples.html), you can see that neither the model or the inputs need to be cast in half precision. That's why you see this `dtype` for the model parameters and inputs.
As for the outputs, this is because the loss computation almost always needs to happen in float32, so we actively convert the outputs to float32 ([here](https://github.com/huggingface/accelerate/blob/517cbf408bf18b7fc9c3dfb9cfb47378d6d31cce/src/accelerate/accelerator.py#L495)). It shouldn't hurt performance.

**Comment by edgarriba** [comment]
@sgugger thanks for the feedback -- I'll try soon in project i'm working right now.

**Comment by roadmap-study** [comment]
> Hi Edgar! Glad you love the library :-)
> 
> I think your expectations come from the old days of mixed precision with apex. With the new PyTorch AMP, and following [their tutorial](https://pytorch.org/docs/stable/notes/amp_examples.html), you can see that neither the model or the inputs need to be cast in half precision. That's why you see this `dtype` for the model parameters and inputs. As for the outputs, this is because the loss computation almost always needs to happen in float32, so we actively convert the outputs to float32 ([here](https://github.com/huggingface/accelerate/blob/517cbf408bf18b7fc9c3dfb9cfb47378d6d31cce/src/accelerate/accelerator.py#L495)). It shouldn't hurt performance.

Hi, 
Does this mean that even though we see the dtype is float32, the actual type is float16?

**Comment by achillesliu** [comment]
> Hi Edgar! Glad you love the library :-)
> 
> I think your expectations come from the old days of mixed precision with apex. With the new PyTorch AMP, and following [their tutorial](https://pytorch.org/docs/stable/notes/amp_examples.html), you can see that neither the model or the inputs need to be cast in half precision. That's why you see this `dtype` for the model parameters and inputs. As for the outputs, this is because the loss computation almost always needs to happen in float32, so we actively convert the outputs to float32 ([here](https://github.com/huggingface/accelerate/blob/517cbf408bf18b7fc9c3dfb9cfb47378d6d31cce/src/accelerate/accelerator.py#L495)). It shouldn't hurt performance.

But how should I deal with it when I want to manually use fp16 for like memory issue?

**Comment by muellerzr** [comment]
You don't want to do this. Check out this documentation here for why: https://huggingface.co/docs/accelerate/concept_guides/fsdp_and_deepspeed#on-differences-in-data-precision-handling
